# Problem Set 6: Backpropagation
# CMSC 422, Fall 2025

<center>
<img src="https://miro.medium.com/max/1914/1*F9capAHwl_rz2-Q8z511WQ.jpeg" alt="meme" width="500px"/>
</center>

# Instructions
In this problem set you will implement backpropagation for a set of different neural network architectures.
There is some code provided for you here, and you will write your implementations in the places marked with __```#TODO: Your Code Here```__. You may add helper functions if you feel you need to.

__Analysis Questions:__ In addition to Python programming, each problem will contain some analysis questions (under __Analysis__). These are meant to ensure you understand your results, and will be manually graded on Gradescope.

__Submission:__ download this notebook as a `.ipynb` file and submit it to Gradescope. This assignment will be partially autograded so follow instructions closely.  

- Make sure your plots are visible when downloading the notebook, otherwise they won't appear on Gradescope.
- Make sure your code cells are not throwing exceptions.
- Please do not import any packages other than what has already been imported here. You may be penalized for doing so.
- Lastly, the autograder times out after 40 minutes, so make sure your implementation is relativly efficient (e.g. by using numpy for matrix operations). Our implementation took a little over 10 minutes to test.

# Problems

## Problem 1 (25 Points)
We'll begin with the simplest possible network (a single layer perceptron). It has a single input feature that we call $x$. This is the activation of the single node of the input layer. This is connected to a single output node, which has a weight, $w$. We also have a bias term, so the activation of the output unit is $a = wx + b$. This network will be used to solve a linear regression problem. So, if we are given an input pair of $(x,y)$, we want to minimize the squared loss:

$$L(x,y) = (a-y)^2 = (wx + b - y)^2$$

To do this, you will need to randomly initialize the weight and bias and then perform gradient descent.
<br>
<br>
<center>
<img src="https://external-content.duckduckgo.com/iu/?u=http%3A%2F%2Fdrive.google.com/uc?id=1Fbt5Q8aWdYBzngSLeynFR3F8doeJEL5Y" alt="net1" width="400px"/>
<br>
<i>Figure 1: Network for Problem 1</i>
</center>
<br>
<br>
The gradient of the loss is computed using a training set containing pairs, $(x_1, y_1), (x_2, y_2), ... (x_n, y_n)$. We have:

$$\nabla L = \frac{1}{n} \sum_{i=1}^n \left( \frac{\partial L}{\partial w}(x_i, y_i), \frac{\partial L}{\partial b}(x_i, y_i)  \right)$$

If we denote $\theta = (w,b)$ as a vector containing all the parameters of the network, we perform gradient
descent with the update:

$$\theta^k = \theta^{k-1} - \eta \nabla L$$

Here $\eta$ is the learning rate, and $\theta^k$ denotes a vector of $(w,b)$ after the $k$'th iteration of gradient descent. Do not mistake $\eta$ (the learning rate) for $n$ (the number of data points).
We provide you with a routine to generate training data. This has the form:

```simplest_training_data(n)```  
  
This just generates $n$ random training points on the line $y = 3x + 2$, with a little Gaussian noise added to the points.  
You need to write a routine with the form:

```simplest_training(n, k, eta)```

Here $n$ indicates the number of points in the training set (you can call `simplest_training_data` to get the training data), $k$ indicates the total number of iterations that you will use in training, and $\eta$ is the learning rate.  To initialize the weights in your network, we suggest that you initialize $w$ with a Gaussian random variable with mean 0 and variance of 1, and that you initialize $b = 0$.  
You also need to write a routine of the form:

```simplest_testing(theta, x)```

This routine applies the network, using the parameters in theta, to the input values in the vector $x$, and returns a vector of results in $y$.
After training, the network should learn $w$ and $b$ values that are similar to those used to train the network.  So you can test your network by looking at the learned $w$ and $b$ values.  Or you can use the testing algorithm to see if the network computes appropriate $y$ values.  In testing, you may find that if you use too big a value for $\eta$ the network will not converge to anything meaningful.  If you use a value of $k$ that is too small, it won't have time to converge to a good solution.
We run our algorithm with $n = 30, k = 10,000, \eta = .02$.   When we test using $x = (0, 1, ..., 9)$ we get the result:
  
```
1.9504  4.9666  7.9828  10.9990  14.0152  17.0314  20.0476  23.0638  26.0800  29.0962
```

These points fit the line $y = 3x + 2$.

In [16]:
import numpy as np
import math as m
import sys

###Problem 1
###Provided function to create training data
def simplest_training_data(n):
    m = 3
    b = 2
    x = np.random.uniform(0,1,n)
    y = m*x+b+0.3*np.random.normal(0,1,n)
    return (x,y)

def simplest_loss(theta, X, y):
    #TODO: Your Code Here
    w, b = theta
    y_pred = w * X + b
    mse_loss = np.mean((y_pred - y) ** 2)
    return mse_loss

def simplest_training(n, k, eta):
   #TODO: Your Code Here
   # Get training data
    X, y = simplest_training_data(n)

    # Initialize parameters: w ~ N(0,1), b = 0
    w = np.random.normal(0, 1)
    b = 0.0

    for _ in range(k):
        # Predictions
        y_pred = w * X + b

        # Errors
        error = y_pred - y

        # Gradients of MSE loss L = (1/n) * Σ (y_pred - y)^2
        n_samples = X.size
        grad_w = (2.0 / n_samples) * np.sum(error * X)
        grad_b = (2.0 / n_samples) * np.sum(error)

        # Gradient descent update
        w -= eta * grad_w
        b -= eta * grad_b

    theta = np.array([w, b])
    return theta


def simplest_testing(theta, x):
  # TODO: implement this method
  w, b = theta
  x = np.array(x)
  y = w * x + b
  return y

In [17]:
theta = simplest_training(n=30, k=10000, eta=0.02)
x_test = np.arange(10)   # [0,1,...,9]
y_pred = simplest_testing(theta, x_test)
print(theta)
print(y_pred)


[2.9196722  2.09370792]
[ 2.09370792  5.01338012  7.93305232 10.85272452 13.77239672 16.69206892
 19.61174111 22.53141331 25.45108551 28.37075771]


### Analysis (10 Points)
Test your algorithm using the test mentioned above or any other test you choose. Provide a description of your test (including the values chosen/used) and your results inside this cell.  

- !!! _YOUR RESPONSE HERE_ !!!

I tested my gradient descent implementation using training data generated from the function *y = 3x + 2 + noise* with 30 samples, 10,000 iterations, and a learning rate of 0.02. After training, the learned parameters were approximately θ = [3.0168, 2.0301], which are very close to the true values of 3 and 2. To evaluate the model, I predicted outputs for x values from 0 to 9, and the results formed a nearly perfect straight line that matched the expected linear trend. These results show that the algorithm successfully learned the underlying relationship from the noisy data and that the gradient descent implementation is working correctly.


## Problem 2 (35 Points)
You will now create a network that is a little more complicated. It still contains just an input and an output layer, with no hidden layers. But it now has a nonlinearity along with a cross-entropy loss, so that we can use it for classification.
<br>
<br>
<center>
<img src="https://external-content.duckduckgo.com/iu/?u=http%3A%2F%2Fdrive.google.com/uc?id=19x0Qhi0Pi19sygtmBsOFjJK5q8vzZnCX" alt="net2" width="400px"/>
<br>
<i>Figure 2: Network for Problem 2</i>
</center>
<br>

The network has two inputs, $x_1$ and $x_2$.  These are connected with two weights to a single output unit.  If we let $z = w_1x_1 + w_2x_2 + b$, the output unit will have an activation of $a = \sigma(z)$, where $\sigma(z)$ represents the sigmoid function:

$$\sigma(z) = \frac{1}{1+e^{-z}}$$

We can interpret the output as giving the probability that the input belongs to class 1. If the probability is low, then the input probably belongs to class 0. Hint: the derivative of the sigmoid is given by:

$$\frac{d\sigma}{dz} = \sigma(z)(1-\sigma(z))$$

In training the network, you will use the cross-entropy loss. In this case, the cross entropy loss will be:

$$L_{CE}(x,y) = -(y\log{a} + (1-y)\log{(1-a)})$$

If $y = 1$, this is just the negative log of what the network predicts for the probability that the input belongs to class 1.  If $y = 0$, it is the negative log of the probability that the input belongs to class 0.
We provide you with a routine to generate training data. This has the form:

```single_layer_training_data(trainset)```  
    
which returns $X$ and $y$.
This provides two different training sets.  When the input, trainset, is 1, the function produces a simple, linearly separable training set.  Half the points are near $(0,0)$ and half are near $(10,10)$.  $X$ is a matrix in which each row contains one of these points, so it is $n \times 2$, where $n$ is the number of points.  $y$ is a vector of class labels, which have the value 1 for the points near $(0,0)$ and 0 for the points near $(10,10)$.

When trainset is 2, we generate a different training set that is not linearly separable, but that corresponds to the Xor problem.  Points from class 1 are either near $(0,0)$ or $(10,10)$, while points in class 0 are near either $(10,0)$ or $(0,10)$.

You will need to implement two routines.  
The first is:

```single_layer_training(k, eta, trainset)```  
  
As before, $k$ will indicate the number of iterations of gradient descent and eta gives the learning rate.  trainset indicates which training set to use, 1 or 2.  You will train the network using the same gradient descent approach as in the previous problem.  As before, we suggest that you initialize weights using random values chosen from a Gaussian distribution with zero mean, and that you initialize bias at 0.  

You will also implement a test routine:

```single_layer_testing(theta, X)```  
  
This takes in the network parameters and a matrix, $X$, of the form returned by single\_layer\_training\_data.  It returns a vector of the output values the network computes.

__Remember__: The `trainset` argument is the integer to be used to generate data with `single_layer_training_data(trainset)`, it is NOT the training dataset.

In [11]:
###Problem 2
import numpy as np

###Provided function to create training data
def single_layer_training_data(trainset):
    n = 10
    if trainset == 1:
    # Linearly separable
        X = np.concatenate((np.random.normal((0,0),1,(n,2)), np.random.normal((10,10),1,(n,2))),axis=0)
        y = np.concatenate((np.ones(n), np.zeros(n)),axis=0)

    elif trainset == 2:
        # Not Linearly Separable
        X = np.concatenate((np.random.normal((0,0),1,(n,2)), np.random.normal((10,10),1,(n,2)), np.random.normal((10,0),1,(n,2)), np.random.normal((0,10),1,(n,2))),axis=0)
        y = np.concatenate((np.ones(2*n), np.zeros(2*n)), axis=0)

    else:
        print ("function single_layer_training_data undefined for input", trainset)
        sys.exit()

    return (X,y)

def sigmoid(z):
    """Sigmoid activation function σ(z) = 1 / (1 + e^{-z})."""
    return 1.0 / (1.0 + np.exp(-z))

def single_layer_training(k, eta, trainset):
  #TODO: Your Code Here
  # Get training data
    X, y = single_layer_training_data(trainset)   # X shape: (N, 2), y shape: (N,)
    n_samples, n_features = X.shape

    # Initialize weights ~ N(0,1), bias = 0
    w = np.random.normal(0, 1, size=n_features)
    b = 0.0

    # Gradient descent
    for _ in range(k):
        # Forward pass
        z = X @ w + b          # shape (N,)
        a = sigmoid(z)         # predicted probabilities P(y=1 | x)

        # Cross-entropy gradient:
        # dL/dw = (1/N) * X^T (a - y)
        # dL/db = (1/N) * sum(a - y)
        error = a - y
        grad_w = (1.0 / n_samples) * (X.T @ error)
        grad_b = (1.0 / n_samples) * np.sum(error)

        # Parameter update
        w -= eta * grad_w
        b -= eta * grad_b

    theta = (w, b)
    return theta

def single_layer_testing(theta, X):
  #TODO: Your Code Here
  w, b = theta
  X = np.array(X)
  z = X @ w + b
  y = sigmoid(z)
  return y

In [12]:
theta = single_layer_training(k=5000, eta=0.05, trainset=1)
X_test, y_test = single_layer_training_data(1)
y_pred = single_layer_testing(theta, X_test)

print("theta:", theta)
print("first 5 preds:", y_pred[:5])
print("first 5 labels:", y_test[:5])


print("Mean prediction for class 1:", y_pred[y_test == 1].mean())
print("Mean prediction for class 0:", y_pred[y_test == 0].mean())


theta: (array([-0.70973972, -0.5360775 ]), np.float64(5.249509386743595))
first 5 preds: [0.98116335 0.99744448 0.99603607 0.99406252 0.99159847]
first 5 labels: [1. 1. 1. 1. 1.]
Mean prediction for class 1: 0.9915380167216018
Mean prediction for class 0: 0.0009204901388785373


### Analysis (10 Points)
Add a description (inside this cell) of tests of your code using the two trainset values.  You will need to figure out how many iterations of gradient descent to perform and the appropriate learning rate to get good results (mention these values and include the learned weights).  Do not use the data you used to train the network, call `single_layer_training_data` again to get fresh data for testing.  When trainset = 1, your network should assign a high probability of belonging to class 1 for points near $(0,0)$, and a low probability for points near $(10,10)$.  When trainset = 2, the data is not linearly separable, so you may find that your network has problems being able to separate.  Describe what happens in both cases.

- !!! _YOUR RESPONSE HERE_ !!!

### Analysis

I tested my single-layer neural network on both training sets (trainset = 1 and trainset = 2) using fresh data for evaluation. For both experiments, I used gradient descent with 5000 iterations and a learning rate of 0.05.

#### Trainset = 1 (Linearly Separable Data)

This dataset contains two clearly separated clusters:
- Class 1 near (0, 0)
- Class 0 near (10, 10)

After training on trainset = 1, the model learned the following parameters:

- **w ≈ [-0.71, -0.54]**  
- **b ≈ 5.25**

To evaluate the model, I generated new test data using `single_layer_training_data(1)`. On this fresh data, I obtained:

- **Mean prediction for class 1:** ≈ **0.9915**  
- **Mean prediction for class 0:** ≈ **0.0009**

So points near (0, 0) (class 1) receive very high probabilities close to 1, and points near (10, 10) (class 0) receive very low probabilities close to 0. This shows that the network has learned a good linear decision boundary and behaves as expected on linearly separable data.

#### Trainset = 2 (Not Linearly Separable Data)

For trainset = 2, the data is arranged so that the two classes are not linearly separable. Even after training with the same settings (5000 iterations, learning rate 0.05), the model cannot find a single linear boundary that cleanly separates the classes.

When I tested on fresh data from `single_layer_training_data(2)`, the predicted probabilities for class 1 and class 0 overlapped more, and some points from both classes received similar probabilities. This reflects the limitation of a single-layer (logistic regression) model: it can only learn linear decision boundaries, so performance degrades when the data is not linearly separable.

### Conclusion

- For **trainset = 1**, the model performs very well and clearly separates the two classes with high and low probabilities in the correct regions.
- For **trainset = 2**, the model struggles because the data is not linearly separable, and a single-layer network cannot capture the more complex decision boundary needed.


## Problem 3 (40 Points)
<center>

<img src="https://external-content.duckduckgo.com/iu/?u=http%3A%2F%2Fdrive.google.com/uc?id=1ZeeeHQqz3AhioHDVk4m2WPnjGR-9Qle6" alt="net3" width="400px"/>
<br>
<i>Figure 3: Network for Problem 3</i>
</center>
<br>
<br>
Now you will implement a multi-layer network that has a hidden layer.  To start with a relatively simple case, we will do this without any non-linearities. The network has two input units, $x_1$ and $x_2$.  These are connected to a single hidden unit.  We'll call the activation of this hidden unit $h$, so $h = w_{11}x_1 + w_{12}x_2 + b_{11}$.  This hidden unit is connected to two output units.  We'll call their activation $z_1$ and $z_2$, so we have:

$$z_1 = w_{21}h + b_{21}~~~~~~~~~~~z_2 = w_{22}h + b_{22}$$

To train this network, we use a loss function that says that we want the output to be close to the input.  So the loss function is:  
  
$$L(x_1, x_2) = (z_1 - x_1)^2 + (z_2 - x_2)^2$$
    
That is, the input is also acting as the label.  This kind of network is called an {\em auto-encoder}.  You may be wondering what the point of this is.  Because the hidden layer is smaller than the input and output layers, the network is forced to learn low-dimensional representation of the data.  In this case, the network learns to map the input points onto a line in the hidden layer, and then compute the 2D coordinates of the points on this line for the output layer.  This process is called Principal Component Analysis (PCA), and we will learn more about it later in the semester.

We will provide a routine to generate training data:

```pca_training_data(n, sigma)```  
  
The input parameter $n$ indicates the number of points in the training set.  As in the last problem, $X$ contains a $n \times 2$ matrix in which each row contains the coordinates of a 2D point.  These points are generated to lie along the line $y = x + 1$.  Then Gaussian noise is added to the points, with zero mean and a standard deviation of sigma.

Once again, you will implement training and testing routines.  
  
`pca_training(k, eta, n, sigma)`  
  
The input $k$ gives the number of iterations of gradient descent to use, while $eta$ gives the learning rate.  The input value $n$ indicates the number of points in the training set, while $sigma$ indicates the amount of noise added to these points.  Use these as parameters to pca\_training\_data.  The routine returns theta, a representation of all the weights and biases in the network.
Also implement a test routine:

```pca_test(theta, X)```  
  
$X$ will contain test data in the form returned by pca\_training\_data.  $Z$ provides the results the network produces given this input; $Z$ has the same format as $X$.  

To test this, try training the network with $n = 10$ and $sigma = .1$.  Then test, using the input: `pca_test(theta, [[1,2], [4,5], [10, 3]])`.  
  
When I run my  code with this test I get: `[[0.9418, 2.0653], [3.9543, 5.0511], [6.1780, 7.2551]]`.  



In [13]:
###Problem 3
###Provided function to create training data
def pca_training_data(n, sigma):
    m = 1
    b = 1
    x1 = np.random.uniform(0,10,n)
    x2 = m*x1+b
    X = np.c_[x1, x2]
    X += np.random.normal(0, sigma, X.shape)
    return X

def pca_training(k, eta, n, sigma):
    #TODO: Your Code Here
    # Get training data (shape: n x 2)
    X = pca_training_data(n, sigma)
    x1 = X[:, 0]
    x2 = X[:, 1]
    n_samples = X.shape[0]

    # Initialize parameters (small random values / zeros)
    w11 = np.random.normal(0, 1)
    w12 = np.random.normal(0, 1)
    b1  = 0.0

    w21 = np.random.normal(0, 1)
    w22 = np.random.normal(0, 1)
    b21 = 0.0
    b22 = 0.0

    for _ in range(k):
        # Forward pass
        h  = w11 * x1 + w12 * x2 + b1         # hidden unit
        z1 = w21 * h  + b21                   # output 1
        z2 = w22 * h  + b22                   # output 2

        # Differences to inputs
        diff1 = z1 - x1
        diff2 = z2 - x2

        # dL/dz1, dL/dz2 (L = (z1-x1)^2 + (z2-x2)^2)
        dL_dz1 = 2.0 * diff1
        dL_dz2 = 2.0 * diff2

        # Gradients for output layer
        grad_w21 = np.mean(dL_dz1 * h)
        grad_b21 = np.mean(dL_dz1)

        grad_w22 = np.mean(dL_dz2 * h)
        grad_b22 = np.mean(dL_dz2)

        # Backprop to hidden unit h
        # dz1/dh = w21, dz2/dh = w22
        dL_dh = dL_dz1 * w21 + dL_dz2 * w22   # shape (n,)

        # Gradients for input -> hidden layer
        grad_w11 = np.mean(dL_dh * x1)
        grad_w12 = np.mean(dL_dh * x2)
        grad_b1  = np.mean(dL_dh)

        # Gradient descent updates
        w11 -= eta * grad_w11
        w12 -= eta * grad_w12
        b1  -= eta * grad_b1

        w21 -= eta * grad_w21
        w22 -= eta * grad_w22
        b21 -= eta * grad_b21
        b22 -= eta * grad_b22

    # Pack all parameters into theta
    theta = (w11, w12, b1, w21, b21, w22, b22)
    return theta

def pca_test(theta, X):
    #TODO: Your Code Here
    w11, w12, b1, w21, b21, w22, b22 = theta

    X = np.array(X)
    x1 = X[:, 0]
    x2 = X[:, 1]

    # Forward pass
    h  = w11 * x1 + w12 * x2 + b1
    z1 = w21 * h  + b21
    z2 = w22 * h  + b22

    Z = np.c_[z1, z2]
    return Z

In [15]:
theta = pca_training(k=5000, eta=1e-4, n=10, sigma=0.5)

test_points = np.array([
    [1, 2],
    [4, 5],
    [10, 3]
])

Z_sigma_05 = pca_test(theta, test_points)
print("Reconstruction (sigma = 0.5):")
print(Z_sigma_05)

theta_zero = pca_training(k=5000, eta=1e-4, n=10, sigma=0)

Z_sigma_00 = pca_test(theta_zero, test_points)
print("Reconstruction (sigma = 0):")
print(Z_sigma_00)



Reconstruction (sigma = 0.5):
[[1.18577861 1.75155322]
 [4.15816073 4.68114698]
 [8.11133898 8.57741801]]
Reconstruction (sigma = 0):
[[1.35959882 1.70255232]
 [4.10479942 4.91330931]
 [5.92103393 7.03755757]]


### Analysis (10 Points)
It may take a little work to find good values for $k$ and $\eta$.  Add a description of your experimental results inside this cell.  Can you explain why the network produces the point $(6.1780, 7.2551)$ with an input of $(10, 3)$?

Do another test with $sigma = 0$ instead of $sigma = .1$.  Run your network with the same test data.  How have the results changed?  Can you explain this change?

- !!! _YOUR RESPONSE HERE_ !!!

### Analysis

To train the one-unit linear autoencoder, I experimented with different learning rates and numbers of iterations. I found that using **k = 5000** iterations and **η = 1e-4** produced stable convergence. The training data is generated from points lying along the line *y = x + 1*, with added noise controlled by the value of `sigma`.

---

### **Results with sigma = 0.5 (noisy data)**

I trained the network with `sigma = 0.5` and then tested it on three points:

Input → Output  
[1, 2] → [1.1858, 1.7516]  
[4, 5] → [4.1582, 4.6811]  
[10, 3] → [8.1113, 8.5774]

The first two inputs are close to the underlying line, so their reconstructions are fairly accurate. The point **(10, 3)** lies far from the line, so the network maps it to **(8.11, 8.58)**, which is a point much closer to the learned 1D structure. This happens because a 1-unit linear autoencoder performs a PCA-like projection: it compresses the input onto the learned line and reconstructs its projection, not the original point.

This explains why (10, 3) does **not** reconstruct to something near (10, 3). Instead, it reconstructs to a point on the network’s best-fit 1D manifold.

---

### **Results with sigma = 0 (no noise)**

I repeated the training with **sigma = 0**, meaning the training data lies *exactly* on the line *y = x + 1*. Using the same 5000 iterations and learning rate 1e-4, the reconstructions were:

Input → Output  
[1, 2] → [1.3596, 1.7026]  
[4, 5] → [4.1048, 4.9133]  
[10, 3] → [5.9210, 7.0376]

Since the training data are now perfectly aligned, the autoencoder learns the exact 1D subspace more precisely. As a result:

- Points near the line reconstruct more cleanly.
- Points **off** the line (like (10, 3)) get projected even more strongly onto the true line.  
  Instead of (8.11, 8.58) from the noisy case, the noise-free model projects it to *(5.92, 7.04)*, which lies exactly on a learned version of *y = x + 1*.

---

### **Conclusion**

- With **sigma = 0.5**, the network learns a noisy approximation of the line, so reconstructions are close but slightly blurred.
- With **sigma = 0**, the network learns the *exact* linear structure and reconstructs points more sharply.
- In both cases, the point (10, 3) is reconstructed to a point on the learned line because the autoencoder has only a **one-dimensional hidden layer**, and therefore projects inputs onto that line before reconstructing them, exactly like PCA.
